src/Algebra/Graph.hs

```diff
@@ -715,44 +768,58 @@ adjacencyList = AM.adjacencyList . toAdjacencyMap
 -- Convert a graph to 'AM.AdjacencyMap'.
 toAdjacencyMap :: Ord a => Graph a -> AM.AdjacencyMap a
 toAdjacencyMap = foldg AM.empty AM.vertex AM.overlay AM.connect
+{-# INLINE toAdjacencyMap #-}
 
 -- Like @toAdjacencyMap@ but specialised for graphs with vertices of type 'Int'.
 toAdjacencyIntMap :: Graph Int -> AIM.AdjacencyIntMap
 toAdjacencyIntMap = foldg AIM.empty AIM.vertex AIM.overlay AIM.connect
+{-# INLINE toAdjacencyIntMap #-}
 
 -- | The /path/ on a list of vertices.
 -- Complexity: /O(L)/ time, memory and size, where /L/ is the length of the
 -- given list.
 --
+-- Good producer.
```

Comment:
Can `path` and `circuit` be made good consumers of lists?

Reply:
The use of direct pattern-matching is the problem here. You will have to write the two functions with `foldr` for them to be a good consumer of lists. I don't see right now a clear way to do that while still being efficient and elegant.

Reply:
OK, your reply looks like a good TODO comment to me :)